raft/src/main/java/org/apache/kafka/raft/internals/VoterSet.java

```diff
@@ -0,0 +1,327 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.raft.internals;
+
+import java.net.InetSocketAddress;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.message.VotersRecord;
+import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.common.feature.SupportedVersionRange;
+
+/**
+ * A type for representing the set of voters for a topic partition.
+ *
+ * It encapsulates static information like a voter's endpoint and their supported kraft.version.
+ *
+ * It providees functionality for converting to and from {@code VotersRecord} and for converting
+ * from the static configuration.
+ */
+final public class VoterSet {
+    private final Map<Integer, VoterNode> voters;
+
+    VoterSet(Map<Integer, VoterNode> voters) {
+        if (voters.isEmpty()) {
+            throw new IllegalArgumentException("Voters cannot be empty");
+        }
+
+        this.voters = voters;
+    }
+
+    /**
+     * Returns the socket address for a given voter at a given listener.
+     *
+     * @param voter the id of the voter
+     * @param listener the name of the listener
+     * @return the socket address if it exist, otherwise {@code Optional.empty()}
+     */
+    public Optional<InetSocketAddress> voterAddress(int voter, String listener) {
+        return Optional.ofNullable(voters.get(voter))
+            .flatMap(voterNode -> voterNode.address(listener));
+    }
+
+    /**
+     * Returns all of the voter ids.
+     */
+    public Set<Integer> voterIds() {
+        return voters.keySet();
+    }
+
+    /**
+     * Adds a voter to the voter set.
+     *
+     * This object is immutable. A new voter set is returned if the voter was added.
+     *
+     * A new voter can be added to a voter set if its id doesn't already exist in the voter set.
+     *
+     * @param voter the new voter to add
+     * @return a new voter set if the voter was added, otherwise {@code Optional.empty()}
+     */
+    public Optional<VoterSet> addVoter(VoterNode voter) {
+        if (voters.containsKey(voter.id())) {
+            return Optional.empty();
```

Comment:
We don't have to. The important invariant for `VoterSet` is that voter ids are unique. The kraft leader will not allow a replica to join the `VoterSet` if its replica id already exists in the voters set.

Reply:
But we are checking the invariant here. Let me give an example. Let's say that the set of voters is `(0, DirId0), (1, DirId1), (2, DirId2)`.

If `(2, DirId2')` (notice that the replica id is the same but the directory id is different) tries to join the set of voters, the leader will call `VoterSet::addVoter((2, DirId2'))`. This call will return `Optional.empty()`. Meaning that the voter was not added.

I think I understand the source of the confusion. Maybe `addVoter` should return different values if `replica id` already exist vs `(replica id, replica directory id)` already exist. The KIP doesn't distinguish between these two cases:
> DUPLICATE_VOTER - when the request contains a replica id is already in the committed set of voters. Note that this means that duplicate replica ids are not allowed. This is useful to make automatic voter addition safer.

from [AddVoter handling](https://cwiki.apache.org/confluence/display/KAFKA/KIP-853%3A+KRaft+Controller+Membership+Changes#KIP853:KRaftControllerMembershipChanges-Handling.3).

Do you mind if we revisit this when we implement [AddVoter RPC and request handling](https://issues.apache.org/jira/browse/KAFKA-16535)? I mainly added this so that I can write some of the tests in `VoterSetTest`. This code is not used by `raft/src/main`.

Note that to implement the `UpdateVoter` RPC I will probably add an `VoterSet::updateVoter` methods that will implement the invariants of that operations.
> If the replica id tracked doesn't have a replica directory id, update it with the replica directory id provided in the request.

from [UpdateVoter handling](https://cwiki.apache.org/confluence/display/KAFKA/KIP-853%3A+KRaft+Controller+Membership+Changes#KIP853:KRaftControllerMembershipChanges-Handling.5).